### ISSUE TITLE
World-Blacklist for rewards -> Reward Queue

### ISSUE THREAD

**Issue by Platinteufel** [PyvesB/AdvancedAchievements#88]
Hello,

in my adventure-world my players should not recieve rewards from AdvancedAchievements. I've defined them as an excluded world, but some achievements like "onlinetime" can also be reached in this world.
Is it possible that the player will get the reward items as soon as he's not in an excluded world anymore? :)

**Comment by PyvesB** [comment]
Hello Platinteufel,

Indeed, I think that would be possible quite easily for the _PlayedTime_ category, I will take care of that in the next update. ;-)

Cheers,

Pyves

**Comment by Platinteufel** [comment]
Thanks Pyves for your awesome support! ;-)
There are two of these categories: Playtime in hours and login in days. :)

**Comment by PyvesB** [comment]
Regarding what you are asking, do you think PlayedTime should be accounted for (i.e. the statistic, not only the reward) even if a player is in a excluded world? This is the current implementation but I'm not really sure it actually makes sense; a server admin might simply want to exclude the statistic from his creative lobby world for instance.

**Comment by Platinteufel** [comment]
Well, that's a good question. In my opinion it's okay if the players statistics for PlayedTime increase if the player is in my creative-world or plays my adventure. 
Maybe it's a bit confusing for the players, if they spend a lot of time in adventure and the PlayedTime statistic doesn't change. 
For Login in Days the statistic have to count if they login in adventure. I think you can't tell the players that they have to be in a special world while logging in for this achievement, so we need a queue anyways.

**Comment by PyvesB** [comment]
There is a problem though if we let the _PlayedTime_ statistic increase , but delay giving the achievement until the player enters a non excluded world: the player will see the progress bars filling up in the GUI and will get confused as he will not be receiving the achievements. He may end up with several achievements in progress (orange colour) with the progress bars filled up to their maximum but without having received any of them.

I think a clearer approach would be "in an excluded world or in creative mode, **NONE** of the statistics increase"; this is both easier for a user to understand, but also to explain to the players. For the _Connection_ category I would add a check to detect world changes or gamemode changes, in order to increment the statistic if the player connected in an excluded world. Therefore he will only have the connection accounted for if at some point during the day he goes into a non excluded world or switches to non creative mode (if _RestrictCreative_ is set to true).

**Comment by Platinteufel** [comment]
Okay, that seems like a good solution for me and probably also for the other users. :)

**Comment by PyvesB** [comment]
I think the implementation of what you have requested is finished. It would be awesome if you could test it and confirm that it is working fine (I've also done a lot a refactoring/cleaning/optimising lately, so if you sport anything else that doesn't work just tell me). ;-)

**Comment by PyvesB** [comment]
I think the implementation of what you have requested is now finished. If you could test it and see if everything works as expected, that would be awesome! As I have done a lot of refactoring/optimising/cleaning up lately (hence the delays in official updates), please tell me if you spot anything else that is not working correctly. ;-)

**Comment by Platinteufel** [comment]
Alright, could you compile a jar for me for testing the beta version? :)

**Comment by PyvesB** [comment]
Here it is:

[AdvancedAchievements-3.0.5.zip](https://github.com/PyvesB/AdvancedAchievements/files/560763/AdvancedAchievements-3.0.5.zip)

(still named version 3.0.5, will probably switch to  version 3.1 sometime in the future once I've done some additional work)

**Comment by Platinteufel** [comment]
I had not much time today, sorry. Nevertheless I found three little bugs in this version while some short tests.
1) `Distance Elytra` progress bar doesn't seem to work **(0:05)**
2) Same with the `Time Played` progress bar. **(0:10)**
3) The stats for `Time Played` seems to reset after relog (~6% before and 0% after relog) **(0:35)**

Related video: https://youtu.be/n0rtTfVhv30

Cheers ;-)

**Comment by PyvesB** [comment]
Thank you for your valuable feedback!

1) Yep, spotted the mistake in the code!

2) and 3) A part from the colour not switching to orange when you are below one hour of game (will be fixed), they are working fine.... well sort of. 

The idea with the Distance and PlayedTime achievements is that they are retrieved lazily from the database when the scheduled task is run. Therefore, if a player does an /aach list after he has connected, but a scheduled task has not yet run, the statistic is simply not available, and the progress bars show 0. If you would have waited a few seconds more in your video, the plugin would have basically behaved as "hey, Platinteufel is now connected, I'm going to fetch the statistic from the database and keep it in memory until he disconnects, at which point I will write it back". 

I could make the delays smaller, this would not impact performance as I've done quite some optimisations on these functions since the latest official release. A player would have to really rush the /aach list as soon as he connects AND be unlucky to see the progress bars empty. This would be acceptable in my opinion, especially regarding the fact that players don't usually type /aach list as their first action when they connect. 

Unfortunately, there is a problem with this approach: for instance if a player is walking the plugin basically says "okay, player is on foot, don't load minecart or other distance statistics". As long as the player doesn't go into a Minecart, the statistic will not load, and won't show up in the progress bars in the GUI. I'm pretty sure this behaviour is also in current release 3.0.5, I'm surprised nobody has yelled at me about it. :-p

To sum up, I'll still make the statistics load lazily as done now, but if a player enters a relevant category in the GUI and the statistic is not available, I will fetch it from the database immediately at that point.

**Comment by PyvesB** [comment]
I've compiled a new version implementing all that was discussed above. Feel free to give any feedback you wish and even open a new ticket if you find any issue non related to our current discussion. Thanks!

[AdvancedAchievements-3.0.5.zip](https://github.com/PyvesB/AdvancedAchievements/files/567370/AdvancedAchievements-3.0.5.zip)

**Comment by Platinteufel** [comment]
I've repeated my tests and everything seems working fine so far. Thank you for still spending time on improving this awesome plugin! ;-)

**Comment by PyvesB** [comment]
Perfect, in that case I'm closing this ticket, but don't hesitate notify me if you have any other issues. ;-)
